docs/DESIGN.md

```diff
@@ -0,0 +1,116 @@
+# Design
+
+The purpose of this document is to describe how `linear-base` is developed.
+_Please be familiar with the user guide before reading this document._
+
+This document describes
+
+ * the overall architecture of linear base
+ * the general implementation strategy
+ * the ideas behind implementing key parts of linear base
+ * rules and advice on contributing
+
+## Overall Architecture
+
+Linear base is more than a copy of things from [`base`] with some function
+arrows being replaced by linear arrows. Moreover, the goal is __not__ exact
+compliance to base.
+
+Linear base consists of the following:
+
+ * fundamental data structures, functions and classes that arise naturally from
+   wanting to do any linear development (e.g., `Unrestricted` and `Consumable`)
+ * tools ported from [`base`] **and from other critical haskell libraries, like `lens`**
+ * new APIs for using external resources, e.g., file IO in [`System.IO.Resource`]
+ * new tools made possible with linear-types, like an API for mutable arrays
+   that doesn't use complicated monads ([`Data.Array.Mutable.Linear`]).
+
+There is a top-level `Prelude.Linear` that is meant to be imported _unqualified_.
+It does not include functors, monads, applicatives and so on because there are
+multiple sensible ways to give linear arrows to these things. See this [blog
+post] for details. This prelude includes:
+
+  * Linear variants of the standard `Prelude`
+  * The fundamental `Data.Unrestricted` data, functions and classes
+  * The `Num` and related classes and instances in the [`Num`] module
+
+
+## General Implementation Strategy
+
+This is the strategy (that we've followed so far) for developing linear base:
+
+1. If the tool is simple enough that there's only one sensible place to replace
+a function arrow by a linear arrow, do that. Example:
+
+```haskell
+foldr :: (a #-> b #-> b) -> b #-> [a] #-> b
+foldr f z = \case
+  [] -> z
+  x:xs -> f x (foldr f z xs)
+```
+
+Otherwise, we need to implement each sensible variant of the tool and put it in a
+well-named module. This is the case with `Data.Functor`s and `Control.Functor`s
+(recall this [blog post]).
+
+2. If we're adding support for an external resource, we may want a new API. For
+this, we need to think carefully about the protocol of using that external
+resource.  That is, we need to identify a clean set of rules that if we obey,
+we can use the resource to do what we want and not cause system errors. For
+instance, we need to open, use, then close file handles. This protocol is
+obeyed in the API in linear base -- see [examples/Simple/FileIO.hs].
+
+3. The ideas behind new tools that are just now possible with linear types vary
+and each have unique concepts that are not addressed by a general strategy.
+These should be documented in the section below.
```

Comment:
Not really, they can be documented in the relevant module. Unless they are an over-arching concept (it makes sense, I suppose, to document explicit departures from base). Best not to clutter the toplevel document.